amqp_events/celery.py

```diff
@@ -1,16 +1,324 @@
+        it's re-published to new exchange preserving initial routing_key.
+        A queue for each exchange has `message-ttl` set to a power of 2. After
+        message is expired, it is re-routed by broker to `dead-letter-exchange`
+        named `recover`. All queues defined for event handlers are also bound to
+        this exchange with same routing key, thus each message after a retry
+        appears in same incoming queue.
+
+        `events` -> `demo.my_event_queue` -> Celery worker (retry)
+            | (publishes new message on retry)
+            V
+        `demo:retry.N` -> `demo:retry.N`
+            | (message ttl expires )
+            V
+        `recover` -> (routing key) -> `demo.my_event_queue` - > Celery worker
+        """
+        channel = self.broker_connection().default_channel
+        for retry in range(defaults.AMQP_EVENTS_MAX_RETRIES):
+            name = self.get_retry_exchange_name(retry)
+            retry_exchange = Exchange(name=name, type=EXCHANGE_TYPE_FANOUT)
+            retry_queue = Queue(
+                name=name,
+                exchange=retry_exchange,
+                queue_arguments={
+                    X_MESSAGE_TTL: 2 ** retry * 1000,  # ms
+                    X_DEAD_LETTER_EXCHANGE: self.recover_exchange_name,
+                }
+            )
+            retry_queue.declare(channel=channel)
+            retry_queue.maybe_bind(channel=channel)
+
+    def _register_archived_queue(self, **_: Any) -> None:
+        """
+        Registers an exchange and a queue for archived messages.
+        """
+        max_ttl = defaults.AMQP_EVENTS_ARCHIVED_MESSAGE_TTL
+        max_len = defaults.AMQP_EVENTS_ARCHIVED_QUEUE_LENGTH
+        if not (max_ttl or max_len):
+            # archived exchange is disabled
+            return
+        archived = Exchange(name=self.archived_exchange_name,
+                            type=EXCHANGE_TYPE_FANOUT)
+        channel = self.broker_connection().default_channel
+        archived_queue = Queue(
+            name=self.archived_exchange_name,
+            exchange=archived,
+            queue_arguments={
+                X_MESSAGE_TTL: max_ttl,
+                X_MAX_LENGTH: max_len,
+                X_QUEUE_MODE: QUEUE_MODE_LAZY
+            })
+        archived_queue.declare(channel=channel)
+        archived_queue.maybe_bind(channel=channel)
+
+
+class Event:
+    """ Event function wrapper that sends a message to broker"""
+
+    def __init__(self, app: EventsCelery, name: str, func: Callable) -> None:
+        self.app = app
+        self.name = name
+        self.func = func
+
+    def __call__(self, *args: Any, **kwargs: Any) -> str:
+        self.func(*args, **kwargs)
+        s = canvas.Signature(
+            args=args,
+            kwargs=kwargs,
+            task=self.name,
+            app=self.app,
+            task_type=self.app.Task,
+            routing_key=self.name)
+        result: AsyncResult = s.apply_async()
+        return result.id
```

Comment:
Почему здесь не вернуть AsyncResult?

Reply:
Because event is asynchronous by it's nature; also it is usually handled by another service or multiple services, and thus can not have any "result" at all. Giving a hope to fetch an event handling result if crucial, unless it's really needed by someone.
I don't want to push users to handle event results by returning `AsyncResult` instance, but it can be changed in any time with reasonable explanation.